Multi-unit support for Origin.js

cuongdo opened the pull request:
Multi-unit listing support for Origin.js

    This commit adds support and tests for multi-unit listings in the
    marketplace resource.

    The key changes use the outstanding offers on a listing to determine the
    units available for purchase in the marketplace resource. The main new
    function is unitsAvailable().

    These changes slow down various offer-related actions, so to keep DApp
    performance reasonable, it's important to be able to pull listing and
    offer data from the discovery server.

### Checklist:

- [x] Test your work and double-check to confirm that you didn't break anything
- [x] Wrap any new text/strings for translation
- [x] Map any new environment variables with a default value in the Webpack config
- [x] Update any relevant READMEs and [docs](https://github.com/OriginProtocol/origin/tree/master/origin-docs)

cuongdo commented:
This is still missing support for per-unit commissions. I'll add support for that after the holiday weekend. Feedback is still appreciated.

franckc commented on origin-js/src/resources/marketplace.js:
```diff
@@ -206,28 +206,51 @@ export default class Marketplace {
+      if (offer.unitsPurchased > unitsAvailable) {
+        return offers
+      }
+      if (offer.status !== 'created' && offer.status !== 'withdrawn') {
```
Maybe add a TODO about handling disputes/refunds case. If the buyer returns the item(s) maybe that should replenish the inventory ?

cuongdo commented:
Thanks @franckc for the review. I'll work on addressing your comments and adding commission support. I'll let you know when this is ready for another review.

cuongdo commented on origin-js/src/resources/marketplace.js:
```diff
@@ -328,6 +372,18 @@ export default class Marketplace {
   async updateListing(listingId, ipfsData, additionalDeposit = 0, confirmationCallback) {
+    const oldListing = await this.getListing(listingId)
+    if (
+      oldListing.type == 'unit' &&
```
Done. In the future, I feel like there should be a less error-prone way (e.g. a function) for detecting listing type.

cuongdo commented on origin-js/src/resources/marketplace.js:
```diff
@@ -279,12 +322,13 @@ export default class Marketplace {
       }
 
-      if (BigNumber(listingCommision).isGreaterThan(BigNumber(chainOffer.commission))) {
+      if (BigNumber(listingCommission).isGreaterThan(BigNumber(chainOffer.commission))) {
```
I've changed this to `!isEqualTo`, *but* I don't think it's that straightforward, so I've added a TODO. I think there will be other corner cases that will come out of the DApp implementation, and we'll probably want to address the various corner cases at the same time.

cc @sparrowDom as a heads up

franckc commented:
Thanks for integrating my feedback.
Yes totally agreed the commission logic will likely need tune up when implementing the DApp to account for the case where seller does not stake the full commission amount upfront and also to account for possible listing edits.
